Make wxrc independent from GUI support

ardovm opened the pull request:
The wxrc program does not depend on the GUI libraries, but only on the base and XML libraries.

The purpose of this pull request is to allow compilation of the wxrc program when GUI support is not required.

The long-term purpose is to allow better compatibility of wxWidgets with cross compilation enviromnents for embedded systems. The wxrc program is a host tool, and the host (or SDK) may not have GUI support.

ardovm committed:
Make wxrc independent from GUI support

The wxrc program does not depend on the GUI libraries, but only on
the base and XML libraries.

vadz commented:
This seems to do 2 things:

1. Build wxrc when GUI is disabled.
2. Add an option for disabling building wxrc.

I'm perfectly fine with (1) and the rationale makes sense to me, but I'm not actually sure if we need (2), we could just continue always building it unconditionally if XML support is enabled, couldn't we?

If you think we do need the option, I'm afraid some nitpicking is in order:

- It should be called something other than `USE_WXRC`, we use `USE_XXX` for libraries and this is not a library. `WITH_WXRC` perhaps? I don't really know, which is partially why I'd rather avoid the question by not having the option at all, but `USE_WXRC` doesn't seem right.
- Ideally it would remain off by default for the non-GUI builds and on for the GUI ones, as in the vast majority of cases this is what you'd want.

What do you think?

ardovm commented:
Hello vadz, and thank you for the quick reply.

> This seems to do 2 things:
> 
>     1. Build wxrc when GUI is disabled.
>     2. Add an option for disabling building wxrc.
> 
> I'm perfectly fine with (1) and the rationale makes sense to me, but I'm not actually sure if we need (2), we could just continue always building it unconditionally if XML support is enabled, couldn't we?
[...]

You are very right. I'll rework the patch to only do (1).

ardovm committed:
Always build wxrc

This commit should be merged into the previous before merging the
branch.

ardovm commented:
I force-pushed away the first Bitbake and autoconf runs.
The second commit amends the first -- please let us merge them before merging the PR. I thought of keeping them separated for the sake of history.
Then I re-ran Bakefile and autoconf.

ardovm commented:
Please note that, in file `build/bakefiles/wx.bkl`, I left the comment starting with "FIXME" because I did not understand it. I can remove it if it was related to the `USE_XRC` conditional.

vadz commented on build/bakefiles/wx.bkl:
```diff
@@ -47,7 +47,7 @@
         <!-- FIXME: this is dirty hack, better bakefile support for
              conditional and optional subprojects is needed -->
-        <action id="wxrc" cond="USE_XRC=='1'">
+        <action id="wxrc">
```
I'm sorry, but why did `USE_XRC` disappear here? It should remain AFAICS...

ardovm replied:
`USE_XRC` is disabled by the `configure` script if GUI support is disabled.
(in fact, the `--enable-xrc` parameter is not evaluated at all)

I tried patching `configure.in` to allow setting `USE_XRC` in non-GUI builds, but compilation failed... on GUI-related XRC classes.

vadz commented on configure.in:
```diff
@@ -8725,10 +8733,14 @@ for subdir in $SUBDIRS; do
                     if test -d $srcdir/utils/$util ; then
-                        makefiles="utils/$util/src/Makefile.in \
-                                   $makefiles"
+                        if test -d $srcdir/utils/$util/src; then
```
This is a nitpick, but it took me some time to understand what was going on here. I think using `test -f ...path.../Makefile.in` would make this more clear as it would read more naturally: if this makefile exists, then add it, otherwise add that one.

vadz reviewed (commented):
Thanks, but I do wonder whether removing `USE_XRC` was intentional and, if so, why.

Also, just to be clear about the meaning of your comment: would you like this to be squash-merged (losing history of the individual commits) or merged normally?

ardovm commented:
> Thanks, but I do wonder whether removing `USE_XRC` was intentional and, if so, why.

I replied inline: it is disabled by default on non-GUI builds.

> Also, just to be clear about the meaning of your comment: would you like this to be squash-merged (losing history of the individual commits) or merged normally?

I am doing many commits because of tries and errors. I do not think that these tries and errors are worth being recorded forever in wxWidgets' history ;-)
If you like best, I can force-push the branch at the end... if you decide to merge it.
I am refraining from doing it now, so that the evolution of this pull request remains clear. I am just getting rid of accessory commits (Bitbake and autotools) because they are not really interesting.

All of this is IMHO, so if you like best to follow another approach, I can do that.

ardovm committed:
Search for Makefile.in explicitly

To be squashed

vadz commented:
> > Thanks, but I do wonder whether removing `USE_XRC` was intentional and, if so, why.
> 
> I replied inline: it is disabled by default on non-GUI builds.

Oops, sorry, I've somehow confused `USE_XRC` and `USE_XML`. I think we need make building wxrc dependent on `USE_XML`.

Of course, allowing to have `USE_XRC=1` in non-GUI builds would be possible too, but this doesn't make much sense.

> > Also, just to be clear about the meaning of your comment: would you like this to be squash-merged (losing history of the individual commits) or merged normally?
> 
> I am doing many commits because of tries and errors. I do not think that these tries and errors are worth being recorded forever in wxWidgets' history ;-)

OK, so it's meant to be squash-merged at the end.

> If you like best, I can force-push the branch at the end... if you decide to merge it.
> I am refraining from doing it now, so that the evolution of this pull request remains clear. I am just getting rid of accessory commits (Bitbake and autotools) because they are not really interesting.

It's called bakefile BTW, I was wondering what bitbake was for a second when I saw it the first time.

> All of this is IMHO, so if you like best to follow another approach, I can do that.

No, _not_ force pushing is indeed best. Thanks!

ardovm commented:
> > > Thanks, but I do wonder whether removing `USE_XRC` was intentional and, if so, why.
> > 
> > I replied inline: it is disabled by default on non-GUI builds.
> 
> Oops, sorry, I've somehow confused `USE_XRC` and `USE_XML`. I think we need make building wxrc dependent on `USE_XML`.

The `USE_XML` symbol does not seem to appear in any .bkl file across the source tree.

Would you like me to try adding it?

[...]
> > If you like best, I can force-push the branch at the end... if you decide to merge it.
> > I am refraining from doing it now, so that the evolution of this pull request remains clear. I am just getting rid of accessory commits (Bitbake and autotools) because they are not really interesting.
> 
> It's called bakefile BTW, I was wondering what bitbake was for a second when I saw it the first time.

Yes, sorry! I get it wrong most of the time. I am working on BitBake files in parallel to this, so it is very easy for me to mix the two names.

vadz commented:
Yes, sorry for not being clear, I think `USE_XML` should be added -- `wxrc` won't compile without it.

ardovm committed:
Add USE_XML to bakefiles and set wxUSE_XML accordingly

The configure script disables wxUSE_XML when the expat library is
disabled.
Compilation of wxrc depends on USE_XML.

ardovm committed:
Bakefile run

ardovm committed:
Configure script updated

ardovm commented:
It _seems_ to work. Can you think of any sensible tests for this new option?

Moreover: I added it to the Bakefile files and to the configure script, because that is what I can use. I don't know if other build systems need to be adjusted as well. Please give me some pointers, if there is any other work to do.

vadz commented:
Thanks, this looks good to me too, so I'm merging it with some minor changes (rerun `./build/update-setup-h` to update all the other generated `setup.h` files; slightly simplify configure).
